crates/indexer/src/helper.rs

```diff
@@ -0,0 +1,191 @@
+            let (from_address, to_address, amount) = parse_sudt_log_data(data);
+            Ok(GwLog::SudtTransfer {
+                sudt_id,
+                from_address,
+                to_address,
+                amount,
+            })
+        }
+        GW_LOG_SUDT_PAY_FEE => {
+            let sudt_id: u32 = item.account_id().unpack();
+            if data.len() != (1 + 20 + 20 + 16) {
+                return Err(anyhow!("Invalid data length: {}", data.len()));
+            }
+            let (from_address, block_producer_address, amount) = parse_sudt_log_data(data);
+            Ok(GwLog::SudtPayFee {
+                sudt_id,
+                from_address,
+                block_producer_address,
+                amount,
+            })
+        }
+        GW_LOG_POLYJUICE_SYSTEM => {
+            if data.len() != (8 + 8 + 20 + 4) {
+                return Err(anyhow!(
+                    "invalid system log raw data length: {}",
+                    data.len()
+                ));
+            }
+
+            let mut u64_bytes = [0u8; 8];
+            u64_bytes.copy_from_slice(&data[0..8]);
+            let gas_used = u64::from_le_bytes(u64_bytes);
+            u64_bytes.copy_from_slice(&data[8..16]);
+            let cumulative_gas_used = u64::from_le_bytes(u64_bytes);
+
+            let created_address = {
+                let mut buf = [0u8; 20];
+                buf.copy_from_slice(&data[16..36]);
+                buf
+            };
+            let mut u32_bytes = [0u8; 4];
+            u32_bytes.copy_from_slice(&data[36..40]);
+            let status_code = u32::from_le_bytes(u32_bytes);
+            Ok(GwLog::PolyjuiceSystem {
+                gas_used,
+                cumulative_gas_used,
+                created_address,
+                status_code,
+            })
+        }
+        GW_LOG_POLYJUICE_USER => {
+            let mut offset: usize = 0;
+            let mut address = [0u8; 20];
+            address.copy_from_slice(&data[offset..offset + 20]);
+            offset += 20;
+            let mut data_size_bytes = [0u8; 4];
+            data_size_bytes.copy_from_slice(&data[offset..offset + 4]);
+            offset += 4;
+            let data_size: u32 = u32::from_le_bytes(data_size_bytes);
+            let mut log_data = vec![0u8; data_size as usize];
+            log_data.copy_from_slice(&data[offset..offset + (data_size as usize)]);
+            offset += data_size as usize;
+            log::debug!("data_size: {}", data_size);
+
+            let mut topics_count_bytes = [0u8; 4];
+            topics_count_bytes.copy_from_slice(&data[offset..offset + 4]);
+            offset += 4;
+            let topics_count: u32 = u32::from_le_bytes(topics_count_bytes);
+            let mut topics = Vec::new();
+            log::debug!("topics_count: {}", topics_count);
+            for _ in 0..topics_count {
+                let mut topic = [0u8; 32];
+                topic.copy_from_slice(&data[offset..offset + 32]);
```

Comment:
Potential out of bound attacks, please check the offset against data length first

Reply:
updated

Reply:
Solidity does allow up to 4 topics (the maximum supported by the EVM).
> Events are an abstraction of the Ethereum logging/event-watching protocol. Log entries provide the contract’s address, a series of up to four topics and some arbitrary length binary data. 
> https://docs.soliditylang.org/en/latest/abi-spec.html?highlight=log#events